File: samples/matlab/s009_projection_matrix.m
% -----------------------------------------------------------------------
% This file is part of the ASTRA Toolbox
% 
% Copyright: 2010-2015, iMinds-Vision Lab, University of Antwerp
%            2014-2015, CWI, Amsterdam
% License: Open Source under GPLv3
% Contact: astra@uantwerpen.be
% Website: http://sf.net/projects/astra-toolbox
% -----------------------------------------------------------------------

vol_geom = astra_create_vol_geom(256, 256);
proj_geom = astra_create_proj_geom('parallel', 1.0, 384, linspace2(0,pi,180));

% For CPU-based algorithms, a "projector" object specifies the projection
% model used. In this case, we use the "strip" model.
proj_id = astra_create_projector('strip', proj_geom, vol_geom);

% Generate the projection matrix for this projection model.
% This creates a matrix W where entry w_{i,j} corresponds to the
% contribution of volume element j to detector element i.
matrix_id = astra_mex_projector('matrix', proj_id);

% Get the projection matrix as a Matlab sparse matrix.
W = astra_mex_matrix('get', matrix_id);


% Manually use this projection matrix to do a projection:
P = phantom(256)';
s = W * P(:);
s = reshape(s, [proj_geom.DetectorCount size(proj_geom.ProjectionAngles, 2)])';
figure(1), imshow(s,[]);

% Because Matlab's matrices are stored transposed in memory compared to C++,
% reshaping them to a vector doesn't give the right ordering for multiplication
% with W. We have to take the transpose of the input and output to get the same
% results (up to numerical noise) as using the toolbox directly.

% Each row of the projection matrix corresponds to a detector element.
% Detector t for angle p is for row 1 + t + p*proj_geom.DetectorCount.
% Each column corresponds to a volume pixel.
% Pixel (x,y) corresponds to column 1 + x + y*vol_geom.GridColCount.


astra_mex_projector('delete', proj_id);
astra_mex_matrix('delete', matrix_id);

File: samples/python/s009_projection_matrix.py
# -----------------------------------------------------------------------
# Copyright: 2010-2022, imec Vision Lab, University of Antwerp
#            2013-2022, CWI, Amsterdam
#
# Contact: astra@astra-toolbox.com
# Website: http://www.astra-toolbox.com/
#
# This file is part of the ASTRA Toolbox.
#
#
# The ASTRA Toolbox is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# The ASTRA Toolbox is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with the ASTRA Toolbox. If not, see <http://www.gnu.org/licenses/>.
#
# -----------------------------------------------------------------------

import astra
import numpy as np

vol_geom = astra.create_vol_geom(256, 256)
proj_geom = astra.create_proj_geom('parallel', 1.0, 384, np.linspace(0,np.pi,180,False))

# For CPU-based algorithms, a "projector" object specifies the projection
# model used. In this case, we use the "line" model.
proj_id = astra.create_projector('line', proj_geom, vol_geom)

# Generate the projection matrix for this projection model.
# This creates a matrix W where entry w_{i,j} corresponds to the
# contribution of volume element j to detector element i.
matrix_id = astra.projector.matrix(proj_id)

# Get the projection matrix as a Scipy sparse matrix.
W = astra.matrix.get(matrix_id)


# Manually use this projection matrix to do a projection:
phantom_id, P = astra.data2d.shepp_logan(vol_geom)
s = W.dot(P.ravel())
s = np.reshape(s, (len(proj_geom['ProjectionAngles']),proj_geom['DetectorCount']))

import pylab
pylab.gray()
pylab.figure(1)
pylab.imshow(s)
pylab.show()

# Each row of the projection matrix corresponds to a detector element.
# Detector t for angle p is for row 1 + t + p*proj_geom.DetectorCount.
# Each column corresponds to a volume pixel.
# Pixel (x,y) corresponds to column 1 + x + y*vol_geom.GridColCount.


astra.projector.delete(proj_id)
astra.data2d.delete(phantom_id)
astra.matrix.delete(matrix_id)
